File: sdks/csharp/Axora.Logger/Sinks/HttpSink.cs
using System.Net.Http.Headers;
using System.Text;
using System.Text.Json;
using System.Threading.Tasks;

namespace Axora.Logger.Sinks;

public sealed class HttpSink : ISink
{
    private const int DefaultTimeoutMs = 5000;
    private readonly HttpClient _httpClient;
    private readonly Uri _uri;
    private readonly Dictionary<string, string> _headers;
    private readonly int _timeoutMs;

    public HttpSink(string? url = null, IDictionary<string, string>? headers = null, int timeoutMs = DefaultTimeoutMs, HttpClient? httpClient = null)
    {
        var resolvedUrl = string.IsNullOrWhiteSpace(url) ? Environment.GetEnvironmentVariable("AXORA_SINK_URL") : url;
        if (string.IsNullOrWhiteSpace(resolvedUrl))
        {
            throw new ArgumentException("HttpSink requires a url or AXORA_SINK_URL");
        }

        _uri = new Uri(resolvedUrl);
        _timeoutMs = timeoutMs > 0 ? timeoutMs : DefaultTimeoutMs;
        _httpClient = httpClient ?? new HttpClient();
        _headers = new Dictionary<string, string>
        {
            ["Content-Type"] = "application/json",
        };

        if (headers is not null)
        {
            foreach (var (key, value) in headers)
            {
                _headers[key] = value;
            }
        }

        var token = Environment.GetEnvironmentVariable("AXORA_SINK_TOKEN");
        if (!string.IsNullOrWhiteSpace(token) && !_headers.ContainsKey("Authorization"))
        {
            _headers["Authorization"] = $"Bearer {token}";
        }
    }

    public async Task ExportAsync(WideEventPayload @event)
    {
        using var request = new HttpRequestMessage(HttpMethod.Post, _uri)
        {
            Content = new StringContent(JsonSerializer.Serialize(@event), Encoding.UTF8, "application/json"),
        };

        foreach (var (key, value) in _headers)
        {
            if (key.Equals("Content-Type", StringComparison.OrdinalIgnoreCase))
            {
                request.Content!.Headers.ContentType = MediaTypeHeaderValue.Parse(value);
            }
            else if (key.Equals("Authorization", StringComparison.OrdinalIgnoreCase))
            {
                request.Headers.Authorization = AuthenticationHeaderValue.Parse(value);
            }
            else
            {
                request.Headers.TryAddWithoutValidation(key, value);
            }
        }

        using var cts = new CancellationTokenSource(_timeoutMs);
        using var response = await _httpClient.SendAsync(request, cts.Token).ConfigureAwait(false);
        if (!response.IsSuccessStatusCode)
        {
            throw new HttpRequestException($"HTTP {(int)response.StatusCode} when exporting AXORA Wide Event");
        }
    }
}

File: sdks/java/src/main/java/com/openakta/logger/sinks/HttpSink.java
package com.openakta.logger.sinks;

import com.openakta.logger.WideEventPayload;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import java.io.IOException;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.Duration;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionException;

public class HttpSink implements Sink {
  private static final ObjectMapper MAPPER = new ObjectMapper();
  private static final int DEFAULT_TIMEOUT_MS = 5000;

  private final HttpClient httpClient;
  private final URI uri;
  private final Map<String, String> headers;
  private final int timeoutMs;

  public HttpSink(String url) {
    this(url, Map.of(), DEFAULT_TIMEOUT_MS, HttpClient.newHttpClient());
  }

  public HttpSink(String url, Map<String, String> headers, int timeoutMs, HttpClient httpClient) {
    String resolvedUrl = url != null && !url.isBlank() ? url : System.getenv("OPENAKTA_SINK_URL");
    if (resolvedUrl == null || resolvedUrl.isBlank()) {
      throw new IllegalArgumentException("HttpSink requires a url or OPENAKTA_SINK_URL");
    }

    this.uri = URI.create(resolvedUrl);
    this.timeoutMs = timeoutMs > 0 ? timeoutMs : DEFAULT_TIMEOUT_MS;
    this.httpClient = httpClient;
    this.headers = new LinkedHashMap<>();
    this.headers.put("Content-Type", "application/json");
    if (headers != null) {
      this.headers.putAll(headers);
    }

    String token = System.getenv("OPENAKTA_SINK_TOKEN");
    if (token != null && !token.isBlank() && !this.headers.containsKey("Authorization")) {
      this.headers.put("Authorization", "Bearer " + token);
    }
  }

  @Override
  public CompletableFuture<Void> export(WideEventPayload event) {
    try {
      HttpRequest.Builder builder =
          HttpRequest.newBuilder(uri)
              .timeout(Duration.ofMillis(timeoutMs))
              .POST(HttpRequest.BodyPublishers.ofString(MAPPER.writeValueAsString(event)));
      headers.forEach(builder::header);

      return httpClient
          .sendAsync(builder.build(), HttpResponse.BodyHandlers.discarding())
          .thenApply(
              response -> {
                if (response.statusCode() >= 400) {
                  throw new CompletionException(
                      new IOException(
                          "HTTP " + response.statusCode() + " when exporting OPENAKTA Wide Event"));
                }
                return null;
              });
    } catch (JsonProcessingException ex) {
      return CompletableFuture.failedFuture(ex);
    }
  }
}
